### PR TITLE
Initial algorithms

### PR DESCRIPTION
**Pull request by ErdemT09**
Here are some initial python algorithms that are prepared for this repository.

### CODE REVIEW

File: leetcode/curated170/medium/paint_house.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import random
+
+def minCost(costs: list[list[int]]) -> int:
+        red, blue, green = 0,0,0
+        for r, b, g in costs:
+            red, blue, green = min(blue, green) + r, min(red, green) + b, min(red, blue) + g
+        return min(red, blue, green)
+
```
Comment (altayhunoglu) [inline leetcode/curated170/medium/paint_house.py:8]: It is quite fine and working faster than most of the other solutions.

image

Can we add a class name to make it suitable for LeetCode Python3 default code?

```
class Solution:
        def minCost(self, costs: List[List[int]]) -> int:
            red, blue, green = 0,0,0
            for r, b, g in costs:
                red, blue, green = min(blue, green) + r, min(red, green) + b, min(red, blue) + g
            return min(red, blue, green)
```

File: leetcode/curated170/easy/armstrong_number.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    k = len(str(n))
+    return get_sum_of_kth_power_digits(n, k) == n
+
+
+if __name__ == "__main__":
+    print(armstrong_number(153))  # True
+    print(armstrong_number(370))  # True
+    print(armstrong_number(1634))  # True
+    print(armstrong_number(85))  # False
+    print(armstrong_number(101))  # False
```
Comment (altayhunoglu) [inline leetcode/curated170/easy/armstrong_number.py:19]: This code is quite fine and beats most of the others:

image

Can we convert it to make it object-oriented?

```
class Solution:
    def get_sum_of_kth_power_digits(self, n, k):
            result = 0
            while n != 0:
                result += (n % 10)**k
                n //= 10
            return result

    def isArmstrong(self, n: int) -> bool:
            k = len(str(n))
            return self.get_sum_of_kth_power_digits(n, k) == n   

if __name__ == '__main__':
     solution = Solution()
     print(solution.isArmstrong(153)) 
     print(solution.isArmstrong(370))  # True
     print(solution.isArmstrong(1634))  # True
     print(solution.isArmstrong(85))  # False
     print(solution.isArmstrong(101))  # False
 

```

File: leetcode/curated170/easy/check_if_a_number_is_majority_element_in_a_sorted_array.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+
+def is_majority_element(nums: list[int], target: int) -> bool:
+    l = len(nums)
+    first_index = first_occur(nums, target,l)
+    minimum_last_index = first_index + l//2
+    if minimum_last_index < l and nums[minimum_last_index] == target:
+        return True
+    return False
+
```
Comment (altayhunoglu) [inline leetcode/curated170/easy/check_if_a_number_is_majority_element_in_a_sorted_array.py:20]: This solution is also quite fine.
 We can convert it to object-oriented:
 ```

 class Solution:
    def isMajorityElement(self, nums: List[int], target: int) -> bool:
            l = len(nums)
            first_index = self.first_occur(nums, target,l)
            minimum_last_index = first_index + l//2
            if minimum_last_index < l and nums[minimum_last_index] == target:
                return True
            return False
        
    def first_occur(self, nums, target, l):
        lo = 0
        hi = l
        while lo < hi:
            mid = lo + (hi-lo)//2
            if nums[mid] < target:
                lo = mid + 1
            elif nums[mid] >= target:
                hi = mid
        return lo
```
 
 Nevertheless, we can also add that solution in which we can employ a very useful function "count":
 
```
 class Solution:
    def isMajorityElement(self, nums: List[int], target: int) -> bool:
        if nums.count(target) > len(nums) // 2:
            return True
        return False
```

Both have almost the same efficiency: ~36 ms.

File: leetcode/curated170/easy/confusing_number.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    map = {6: 9, 9: 6, 0: 0, 1: 1, 8: 8}
+    new_num = 0
+    x = n
+    while x != 0:
+        remainder = x % 10
+        if remainder not in map:
+            return False
+        new_num = new_num*10 + map[remainder]
+        x //= 10
+    return n != new_num
```
Comment (altayhunoglu) [inline leetcode/curated170/easy/confusing_number.py:11]: Quite fine:
image

OOP:
```
class Solution:
    def confusingNumber(self, N: int) -> bool:
        map = {6: 9, 9: 6, 0: 0, 1: 1, 8: 8}
        new_num = 0
        x = N
        while x != 0:
            remainder = x % 10
            if remainder not in map:
                return False
            new_num = new_num*10 + map[remainder]
            x //= 10
        return N != new_num
```

File: leetcode/curated170/easy/count_substrings_with_only_one_distinct_letter.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+def count_letters(s: str) -> int:
+    total = 1
+    count = 1
+    for i in range(1, len(s)):
+        if s[i] == s[i-1]:
+            count += 1
+        else:
+            count = 1
+        total += count
+    return total
```
Comment (altayhunoglu) [inline leetcode/curated170/easy/count_substrings_with_only_one_distinct_letter.py:10]: Quite fine:

image


OOP:
class Solution:
    def countLetters(self, S: str) -> int:
        total = 1
        count = 1
        for i in range(1, len(S)):
            if S[i] == S[i-1]:
                count += 1
            else:
                count = 1
            total += count
        return total

File: leetcode/curated170/easy/design_compressed_string_iterator.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        if not self.hasNext():
+            return ' '
+        if self.num == 0:
+            self.ch = self.res[self.pointer]
+            self.pointer += 1
+            while self.pointer<len(self.res) and self.res[self.pointer].isdigit:
+                self.num = self.num * 10 + self.res[self.pointer] - '0'
+                self.pointer += 1
+        self.num -= 1
+        return self.ch
```
Comment (altayhunoglu) [inline leetcode/curated170/easy/design_compressed_string_iterator.py:17]: We had to fix some parts:

```
class StringIterator:
    res = ""
    pointer, num = 0, 0
    ch = ' '
    
    def __init__(self, s: str):
        self.res = s
    
    def next(self) -> chr:
        tmp_num = ""
        if not self.hasNext():
            return ' '
        if self.num == 0:
            self.ch = self.res[self.pointer]
            self.pointer += 1
            while self.pointer<len(self.res) and self.res[self.pointer].isnumeric():
                tmp_num += self.res[self.pointer]
                self.pointer += 1
            self.num = int(tmp_num)
        self.num -= 1
        return self.ch
    
    def hasNext(self) -> bool:
        return self.pointer != len(self.res) or self.num != 0

```
We can also use this:

```
    def next(self) -> chr:
        tmp_num = ""
        if not self.hasNext():
            return ' '
        if self.num == 0:
            self.ch = self.res[self.pointer]
            self.pointer += 1
            while self.pointer<len(self.res) and self.res[self.pointer].isnumeric():
                tmp_num += self.res[self.pointer]
                self.pointer += 1
            self.num = int(tmp_num)
        self.num -= 1
        return self.ch
```

File: leetcode/curated170/easy/find_anagram_mappings.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    mapping = []
+    for x in A:
+        mapping.append(locations[x])
+    return mapping
+
+
+if __name__ == "__main__":
+    A = [12, 28, 46, 32, 50]
+    B = [50, 12, 32, 46, 28]
+    print(anagram_mappings(A, B))  # [1, 4, 3, 2, 0]
```
Comment (altayhunoglu) [inline leetcode/curated170/easy/find_anagram_mappings.py:14]: Quite fine:

image


OOP:
```
class Solution:
    def anagramMappings(self, A: List[int], B: List[int]) -> List[int]:
      
        locations = {}
        for i in range(len(B)):
            locations[B[i]] = i
        mapping = []
        for x in A:
            mapping.append(locations[x])
        return mapping

 
```

File: leetcode/curated170/medium/nested_list_weight_sum.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    :type nestedList: List[NestedInteger]
+    :rtype: int
+    """
+    ret = 0
+    for element in nestedList:
+        if element.isInteger():
+            ret += currLevel * element.getInteger()
+        else:
+            ret += depthSum(element.getList(), currLevel=currLevel + 1)
+    return ret
```
Comment (altayhunoglu) [inline leetcode/curated170/medium/nested_list_weight_sum.py:12]: Original LeetCode method needs just one parameter for nestedList although we need two parameters for DFS.
It is why we can overcome that problem like this:

```
class Solution:

         def depthSum(self, nestedList: List[NestedInteger]) -> int:
    
             def depthSum(nestedList, currLevel=1) -> int:
                ret = 0
                for element in nestedList:
                    if element.isInteger():
                        ret += currLevel * element.getInteger()
                    else:
                        ret += depthSum(element.getList(), currLevel=currLevel + 1)
                return ret
            
             return depthSum(nestedList, 1)
```

File: leetcode/curated170/medium/paint_house.py (hunk 1 above)
Comment (ErdemT09) [reply leetcode/curated170/medium/paint_house.py:8]: By parameters, it should be `costs: list[list[int]]`. Explicitly writing parameter types are naturally not required in Python.

File: leetcode/curated170/easy/check_if_a_number_is_majority_element_in_a_sorted_array.py (hunk 3 above)
Comment (ErdemT09) [reply leetcode/curated170/easy/check_if_a_number_is_majority_element_in_a_sorted_array.py:20]: Python then probably internally utilizes some binary search. Explicitly showing how this works should also remain in my opinion.

File: leetcode/curated170/medium/paint_house.py (hunk 1 above)
Comment (ErdemT09) [reply leetcode/curated170/medium/paint_house.py:8]: Python's naming convention is actually snake_case. Good way of remembering this is that the animal python is a snake.

File: leetcode/curated170/medium/nested_list_weight_sum.py (hunk 8 above)
Comment (ErdemT09) [reply leetcode/curated170/medium/nested_list_weight_sum.py:12]: Actually, it should work fine with LeetCode. In Python, variables who are given explicit values in method call (like currLevel=1) compile fine even if the call of the method does not include it.

Comment (ErdemT09) [reply leetcode/curated170/medium/nested_list_weight_sum.py:12]: Example:
image

File: leetcode/curated170/medium/paint_house.py (hunk 1 above)
Comment (altayhunoglu) [reply leetcode/curated170/medium/paint_house.py:8]: In Python, as far as I know, it is allowed using camelCase to be compliant with pre-existing code conventions.
I think it is why Leetcode prefers using camelCase. When we use snake_case, we get errors.
Also, it obliges using OOP with the "self" parameter.

image

File: leetcode/curated170/easy/check_if_a_number_is_majority_element_in_a_sorted_array.py (hunk 3 above)
Comment (altayhunoglu) [reply leetcode/curated170/easy/check_if_a_number_is_majority_element_in_a_sorted_array.py:20]: Yeah, you are right. It uses TimSort as far as I know.
Of course, Binary Search is quite fine and should remain as well.

File: leetcode/curated170/medium/nested_list_weight_sum.py (hunk 8 above)
Comment (altayhunoglu) [reply leetcode/curated170/medium/nested_list_weight_sum.py:12]: Yep. You are absolutely right.
It works like that also:

```
class Solution:

        def depthSum(self, nestedList, currLevel=1) -> int:
            """
            :type nestedList: List[NestedInteger]
            :rtype: int
            """
            ret = 0
            for element in nestedList:
                if element.isInteger():
                    ret += currLevel * element.getInteger()
                else:
                    ret += self.depthSum(element.getList(), currLevel=currLevel + 1)
            return ret
```

File: leetcode/curated170/medium/paint_house.py (hunk 1 above)
Comment (ErdemT09) [reply leetcode/curated170/medium/paint_house.py:8]: I didn't think about how LeetCode obliges a certain method name. Understood.

File: leetcode/curated170/medium/nested_list_weight_sum2.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+        return max(depth) + 1 if len(depth) > 0 else 1
+
+    def sumDepth(self, nestedList, depth):
+        s = 0
+        for ni in nestedList:
+            if ni.isInteger():
+                s += ni.getInteger() * depth
+            else:
+                s += self.sumDepth(ni.getList(), depth - 1)
+        return s
```
Comment (altayhunoglu) [inline leetcode/curated170/medium/nested_list_weight_sum2.py:26]: Quite fine:
image

File: leetcode/curated170/medium/sparse_matrix_multiplication.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    ncola = len(a[0])
+    ncolb = len(b[0])
+    # assertion, ncola==nrowb
+    c = [[0] * ncolb for x in range(nrowa)]
+    for i in range(nrowa):
+        for k in range(ncola):
+            if a[i][k] != 0:
+                for j in range(ncolb):
+                    c[i][j] += a[i][k] * b[k][j]
+    return c
```
Comment (altayhunoglu) [inline leetcode/curated170/medium/sparse_matrix_multiplication.py:12]: It is also quite fine.

Can we convert it to OOP too:
image

```
class Solution:
       def multiply(self, a, b):
        nrowa = len(a)
        ncola = len(a[0])
        ncolb = len(b[0])
        # assertion, ncola==nrowb
        c = [[0] * ncolb for x in range(nrowa)]
        for i in range(nrowa):
            for k in range(ncola):
                if a[i][k] != 0:
                    for j in range(ncolb):
                        c[i][j] += a[i][k] * b[k][j]
        return c
```

File: leetcode/curated170/medium/wiggle_sort.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import random
+
+def wiggle_sort(nums: list[int]) -> list[int]:
+    for i in range(len(nums) - 1):
+            if (i % 2 == 0) == (nums[i] > nums[i + 1]):
+                nums[i], nums[i + 1] = nums[i + 1], nums[i]
+    return nums
+
```
Comment (altayhunoglu) [inline leetcode/curated170/medium/wiggle_sort.py:8]: Quite fine:
image


Can we convert to OOP:
```
class Solution:
    def wiggleSort(self, nums: List[int]) -> None:
        for i in range(len(nums) - 1):
                if (i % 2 == 0) == (nums[i] > nums[i + 1]):
                    nums[i], nums[i + 1] = nums[i + 1], nums[i]
        return nums
```

File: math/math_py.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    while a <= b:
+        integral += f(a)*dx
+        a += dx
+    return integral
+
+def inf_sum(f, start = 1, end = 1000):
+    sum = 0
+    for k in range(start,end):
+        sum += f(k)
+    return sum
```
Comment (altayhunoglu) [inline math/math_py.py:14]: It is also quite fine.

File: leetcode/curated170/easy/design_compressed_string_iterator.py (hunk 6 above)
Comment (ErdemT09) [reply leetcode/curated170/easy/design_compressed_string_iterator.py:17]: I will actually add the second one, as Python makes this process a bit more clear.

**Review by altayhunoglu — APPROVED** [review]
Thanks for your valuable efforts. Approved.

**Pull request merged**